enhancements/test-platform/e2e-observer-pods.md

```diff
@@ -0,0 +1,98 @@
+---
+title: e2e-observer-pods
+authors:
+  - "@deads2k"
+reviewers:
+approvers:
+  - "@stevek"
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also:
+replaces:
+superseded-by:
+---
+
+# e2e Observer Pods
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+e2e tests have multiple dimensions:
+ 1. which tests are running (parallel, serial, conformance, operator-specific, etc)
+ 2. which platform (gcp, aws, azure, etc)
+ 3. which configuration of that platform (proxy, fips, ovs, etc)
+ 4. which version is running (4.4, 4.5, 4.6, etc)
+ 5. which version change is running (single upgrade, double upgrade, minor upgrade, etc)
+Our individual jobs are an intersection of these dimensions and that intersectionality drove the development of steps
+to help handle the complexity.
+
+There is a kind of CI test observer that wants to run across all of these intersections by leveraging the uniformity of
+OCP across all the different dimensions above.
+It wants to do something like start an observer agent of some kind for every CI cluster ever created and provide o(100M)
+of data back to include in the CI job overall.
+Working with steps would require integrating in multiple dimensions and lead to gaps in coverage.
```

Comment:
You don't need to make contentious, inflammatory and incorrect statements. There's _one_ base IPI workflow and you change that, you affect _all_ IPI clusters under test. Please tell me that's hard.

Reply:
> You don't need to make contentious, inflammatory and incorrect statements. There's _one_ base IPI workflow and you change that, you affect _all_ IPI clusters under test. Please tell me that's hard.

It doesn't cover UPI clusters, which are also guaranteed to have a kubeconfig, making this data available.  This is a gap in coverage.

I don't think construing feedback about how the step approach leads to gaps and requires integrating across multiple dimensions (both of which you just confirmed) should be taken as inflammatory or incorrect.

Reply:
N=2. IPI, UPI. The definition of the observer can be shared. Do you consider a function definition that is called in two places too onerous and require a system redesign so the function is only called once in the program? No. This is poor feedback, not entirely germane to the proposal (steps, by design, do not and will not run in parallel), and not helpful.

Reply:
UPI/IPI, platform, platform-config, and (somewhat) cluster version are all setup/teardown issues that do not have much impact on the test step/chain itself.  E.g. [openshift-e2e-gcp-upi](https://steps.svc.ci.openshift.org/workflow/openshift-e2e-gcp-upi) and [openshift-e2e-aws-ovn](https://steps.svc.ci.openshift.org/workflow/openshift-e2e-aws-ovn) both use [openshift-e2e-test](https://steps.svc.ci.openshift.org/workflow/reference/openshift-e2e-test).  So leaving out folks who use other test steps for now, you could get this behavior today for the large class of `openshift-e2e-test`-consuming jobs by editing [one file][1].  Why don't we do that, and then come back in a follow-up round if we feel like that approach is going great for the covered jobs and we really miss it on the custom-test jobs?

[1]: https://github.com/openshift/release/blob/1be798c30fb41fcd6cd9608387294c3483f39838/ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh